Add macOS compatibility

ygoumaz opened the pull request:
Currently, `webots_ros2` cannot be used on macOS. This PR aims to add the necessary framework to allow such compatibility.

The solution is as follows:
- The user needs to install Docker on its machine.
- A Docker container based on the ROS2 image (arm64) is created.
- The Docker container shares a folder with the host and contains a special environment variable (`WEBOTS_SHARED_FOLDER`) that allows `webots_ros2` to detect the macOS environment.
- `webots_ros2` must be installed in the container.
- In this case (unlike Linux and Windows), Webots is not launched directly from the package but using the shared folder. A Webots "server" is used to start Webots, with the world and the additional arguments found in the shared volume.
- The connection of the controller(s) is possible through TCP.

ygoumaz commented:
The PR is ready for review. In the meantime, I will add the documentation to the wiki. The webots server to start the simulation on the host is also ready for review here: https://github.com/cyberbotics/webots-server/pull/13.

omichel commented on webots_ros2_driver/webots_ros2_driver/webots_launcher.py:
```diff
@@ -161,7 +194,11 @@ def _shutdown_process(self, context, *, send_sigint):
     def __init__(self, output='screen', respawn=True, **kwargs):
-        controller_url = 'tcp://' + get_wsl_ip_address() + ':1234/' if is_wsl() else ''
+        if is_macOS():
+            tcp_url = "host.docker.internal"
```
```suggestion
            tcp_url = 'host.docker.internal'
```

omichel commented on webots_ros2_universal_robot/launch/robot_launch/robot_nodes_launch.py:
```diff
@@ -35,7 +35,12 @@ def generate_launch_description():
 
-    controller_url = 'tcp://' + get_wsl_ip_address() + ':1234/' if is_wsl() else ''
+    if is_macOS():
+        tcp_url = "host.docker.internal"
```
```suggestion
        tcp_url = 'host.docker.internal'
```

omichel commented on webots_ros2_universal_robot/launch/multirobot_launch.py:
```diff
@@ -43,7 +43,12 @@ def get_ros2_nodes(*args):
 
-    controller_url = 'tcp://' + get_wsl_ip_address() + ':1234/' if is_wsl() else ''
+    if is_macOS():
+        tcp_url = "host.docker.internal"
```
```suggestion
        tcp_url = 'host.docker.internal'
```

omichel commented on webots_ros2_turtlebot/launch/robot_launch.py:
```diff
@@ -40,7 +40,12 @@ def generate_launch_description():
 
-    controller_url = 'tcp://' + get_wsl_ip_address() + ':1234/' if is_wsl() else ''
+    if is_macOS():
+        tcp_url = "host.docker.internal"
```
```suggestion
        tcp_url = 'host.docker.internal'
```

omichel commented on webots_ros2_tiago/launch/robot_launch.py:
```diff
@@ -49,7 +49,12 @@ def generate_launch_description():
 
-    controller_url = 'tcp://' + get_wsl_ip_address() + ':1234/' if is_wsl() else ''
+    if is_macOS():
+        tcp_url = "host.docker.internal"
```
```suggestion
        tcp_url = 'host.docker.internal'
```

omichel commented on webots_ros2_tests/test/test_system_driver.py:
```diff
@@ -51,7 +51,12 @@ def generate_test_description():
 
-    controller_url = 'tcp://' + get_wsl_ip_address() + ':1234/' if is_wsl() else ''
+    if is_macOS():
+        tcp_url = "host.docker.internal"
```
```suggestion
        tcp_url = 'host.docker.internal'
```

omichel commented on webots_ros2_tesla/launch/robot_launch.py:
```diff
@@ -38,7 +38,12 @@ def generate_launch_description():
 
-    controller_url = 'tcp://' + get_wsl_ip_address() + ':1234/' if is_wsl() else ''
+    if is_macOS():
+        tcp_url = "host.docker.internal"
```
```suggestion
        tcp_url = 'host.docker.internal'
```

omichel commented on webots_ros2_mavic/launch/robot_launch.py:
```diff
@@ -38,7 +38,12 @@ def generate_launch_description():
 
-    controller_url = 'tcp://' + get_wsl_ip_address() + ':1234/' if is_wsl() else ''
+    if is_macOS():
+        tcp_url = "host.docker.internal"
```
```suggestion
        tcp_url = 'host.docker.internal'
```

omichel commented on webots_ros2_epuck/launch/robot_launch.py:
```diff
@@ -40,7 +40,12 @@ def generate_launch_description():
 
-    controller_url = 'tcp://' + get_wsl_ip_address() + ':1234/' if is_wsl() else ''
+    if is_macOS():
+        tcp_url = "host.docker.internal"
```
```suggestion
        tcp_url = 'host.docker.internal'
```

omichel commented on webots_ros2_driver/webots_ros2_driver/utils.py:
```diff
@@ -105,7 +106,26 @@ def get_wsl_ip_address():
         file.close()
-    
+
+def is_macOS():
```
I would replace this with:
```suggestion
def has_shared_folder():
```
as it may not be always specific to macOS.

ygoumaz commented:
The [wiki page](https://github.com/cyberbotics/webots_ros2/wiki/Complete-Installation-Guide) is now updated with the instructions to use `webots_ros2` on macOS.

omichel reviewed (changes requested):
I would rename `WEBOTS_SHARED_FOLDERS` to `WEBOTS_SHARED_FOLDER`.
Also, the `FAIL:x` error messaging could be improved: instead of `FAIL:0` we could send `FAIL: Error in clear text`.
Finally, the local server could send notifications when Webots quits, etc. so that it behaves the same as on WSL and Linux.

ygoumaz commented:
The PR is finally ready for review again. I refactored quite a lot the code:

- The connection to the Webots starting server is now done in a python subprocess (`webots_tcp_client.py`). This way, when Webots is closed the subprocess receives a message and exits. The other nodes can be shutdown too.
- The shared folder is now cleaned from the ROS side and not from the server side.
- Failure messages are sent through TCP.
- In the universal robot example, the reset handler (which respawns all nodes when the simulation is reset) must be unregistered when closing webots to avoid additional copy of resource files to the shared folder.
- Bug fixes and cleanups.

ygoumaz commented:
I will now improve the documentation by providing a Dockerfile to install the complete environment.

omichel reviewed (approved):
It seems good to me.

I only have a doubt regarding the way the arguments are passed to the local simulation server.
Currently they are stored in a file in the shared folder with a specific name (and a small risk of name clash).

I was wondering whether it could be better (more generic) to send them to the local simulation server as the first connection message, e.g., the message would mean something like "please start webots with the following command line parameters including the world filename", it could take the following form: `COMMAND:webots --batch --mode=RUN /my/shared/folder/worlds/my_world.wbt`. This would simplify the checks on the server side and would be more generic, e.g., allow to launch any command on the host machine. For security reasons we should however, check that the command, e.g. "webots" is in a list of accepted commands.

ygoumaz commented:
I will apply these changes. You are right that is will be a more generic and clean solution. Meanwhile, the wiki page is now updated.  The Docker installation is now summarized to a Dockerfile to build.

omichel commented on webots_ros2_driver/scripts/webots_tcp_client.py:
```diff
@@ -35,7 +37,8 @@ def host_shared_folder():
     print('WARNING: Unable to start Webots. Please start the local simulation server on your host machine. Next connection attempt in 1 second.', file=sys.stderr)
     time.sleep(1)
-tcp_socket.sendall(host_shared_folder().encode('utf-8'))
+command = 'webots ' + os.path.join(host_shared_folder(), world_name) + ' ' + launch_arguments
```
So we are not passing the full path to the webots binary here?
This assume the local server script should prepend the WEBOTS_HOME to it?
Do we want this?

ygoumaz replied:
As discussed, here we only send 'webots' as executable. The server is responsible to append the WEBOTS_HOME prefix.

omichel reviewed (approved):
Nice.

ygoumaz commented:
The tests are failing, because the last nightly of Webots is not available. We can wait until Monday to see if all tests pass to merge the PR.
